analytical_engine/apps/pregel/louvain/louvain_vertex.h

```diff
@@ -0,0 +1,138 @@
+/** Copyright 2020 Alibaba Group Holding Limited.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef ANALYTICAL_ENGINE_APPS_PREGEL_LOUVAIN_LOUVAIN_VERTEX_H_
+#define ANALYTICAL_ENGINE_APPS_PREGEL_LOUVAIN_LOUVAIN_VERTEX_H_
+
+#include <list>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "core/app/pregel/pregel_vertex.h"
+
+#include "apps/pregel/louvain/louvain_context.h"
+
+namespace gs {
+
+template <typename FRAG_T, typename VD_T, typename MD_T>
+class LouvainVertex : public PregelVertex<FRAG_T, VD_T, MD_T> {
+  using fragment_t = FRAG_T;
+  using oid_t = typename fragment_t::oid_t;
+  using vid_t = typename fragment_t::vid_t;
+  using edata_t = double;
+  using vertex_t = typename fragment_t::vertex_t;
+  using adj_list_t = typename fragment_t::const_adj_list_t;
+  using compute_context_t = PregelComputeContext<fragment_t, VD_T, MD_T>;
+  using context_t = LouvainContext<FRAG_T, compute_context_t>;
+  using state_t = LouvainNodeState<vid_t>;
+
+ public:
+  using vd_t = VD_T;
+  using md_t = MD_T;
+
+  state_t& state() { return context_->GetVertexState(vertex_); }
+
+  vertex_t vertex() const { return vertex_; }
+
+  adj_list_t outgoing_edges() { return fragment_->GetOutgoingAdjList(vertex_); }
```

Comment:
Maybe no need to rewrite it,  because a virtual function already exists in the base(PregelVertex) class .

Reply:
fix.